File: app/python/auth/modules/oauth2_client_credentials.py
from __future__ import annotations

import asyncio
import hashlib
import logging
import os
from typing import Any

import httpx
from redis.asyncio import Redis

from app.python.auth.base import AuthContext, AuthResult, _invalidate_on_from_config

log = logging.getLogger(__name__)

_LOCK_TTL = 30
_LOCK_RETRY_COUNT = 20
_LOCK_RETRY_SLEEP = 0.25


class OAuth2ClientCredentialsAuth:
    """OAuth2 client-credentials flow with Redis token cache and distributed lock."""

    name = "oauth2_client_credentials"

    def __init__(
        self,
        redis: Redis,
        http: httpx.AsyncClient,
        client_id: str,
        client_secret: str,
        token_url: str,
        scope: str,
        audience: str,
        redis_key: str,
        redis_lock_key: str,
        header: str,
        prefix: str,
        invalidate_on: set[int],
        ttl_fallback: int,
    ) -> None:
        self._redis = redis
        self._http = http
        self._client_id = client_id
        self._client_secret = client_secret
        self._token_url = token_url
        self._scope = scope
        self._audience = audience
        self._redis_key = redis_key
        self._redis_lock_key = redis_lock_key
        self._header = header
        self._prefix = prefix
        self._invalidate_on = invalidate_on
        self._ttl_fallback = ttl_fallback

    @classmethod
    def from_config(
        cls, config: dict[str, Any], *, redis: Redis, http: httpx.AsyncClient
    ) -> OAuth2ClientCredentialsAuth:
        cid_env = config["client_id_env"]
        csec_env = config["client_secret_env"]
        client_id = os.environ.get(cid_env)
        if not client_id:
            raise ValueError(
                f"OAuth2ClientCredentialsAuth: env var {cid_env!r} is not set or empty."
            )
        client_secret = os.environ.get(csec_env)
        if not client_secret:
            raise ValueError(
                f"OAuth2ClientCredentialsAuth: env var {csec_env!r} is not set or empty."
            )
        token_url = config["token_url"]
        url_hash = hashlib.md5(token_url.encode()).hexdigest()[:8]
        default_redis_key = f"oauth2:{url_hash}"
        redis_key = config.get("redis_key", default_redis_key)
        redis_lock_key = config.get("redis_lock_key", f"{redis_key}:lock")
        return cls(
            redis=redis,
            http=http,
            client_id=client_id,
            client_secret=client_secret,
            token_url=token_url,
            scope=config.get("scope", ""),
            audience=config.get("audience", ""),
            redis_key=redis_key,
            redis_lock_key=redis_lock_key,
            header=config.get("header", "Authorization"),
            prefix=config.get("prefix", "Bearer "),
            invalidate_on=_invalidate_on_from_config(config),
            ttl_fallback=config.get("ttl_fallback", 3600),
        )

    async def _fetch_token(self) -> tuple[str, int]:
        data: dict[str, str] = {
            "grant_type": "client_credentials",
            "client_id": self._client_id,
            "client_secret": self._client_secret,
        }
        if self._scope:
            data["scope"] = self._scope
        if self._audience:
            data["audience"] = self._audience
        resp = await self._http.post(self._token_url, data=data, timeout=15.0)
        resp.raise_for_status()
        payload = resp.json()
        token: str = payload["access_token"]
        expires_in: int = int(payload.get("expires_in", self._ttl_fallback))
        ttl = max(60, expires_in - 60)
        return token, ttl

    async def apply(self, ctx: AuthContext) -> AuthResult:
        raw = await self._redis.get(self._redis_key)
        if raw:
            token = raw if isinstance(raw, str) else raw.decode()
            return AuthResult(headers={self._header: f"{self._prefix}{token}"})

        got_lock = await self._redis.set(self._redis_lock_key, "1", nx=True, ex=_LOCK_TTL)
        if not got_lock:
            for _ in range(_LOCK_RETRY_COUNT):
                await asyncio.sleep(_LOCK_RETRY_SLEEP)
                raw = await self._redis.get(self._redis_key)
                if raw:
                    token = raw if isinstance(raw, str) else raw.decode()
                    return AuthResult(headers={self._header: f"{self._prefix}{token}"})
            raise RuntimeError(
                f"Timed out waiting for peer to fetch OAuth2 token (key={self._redis_key!r})."
            )

        try:
            token, ttl = await self._fetch_token()
            await self._redis.set(self._redis_key, token, ex=ttl)
            log.info("Fetched OAuth2 token for %s, ttl=%ds", self._token_url, ttl)
            return AuthResult(headers={self._header: f"{self._prefix}{token}"})
        finally:
            await self._redis.delete(self._redis_lock_key)

    async def invalidate(self) -> None:
        await self._redis.delete(self._redis_key)

    def is_rejection(self, status_code: int, body: dict | None) -> bool:
        return status_code in self._invalidate_on

File: app/ts/src/auth/modules/oauth2_client_credentials.ts
import type { AuthContext, AuthResult } from "../../core/types.ts";
import type { AuthModule, AuthDeps } from "../types.ts";
import { invalidateOnFromConfig } from "../types.ts";

const LOCK_TTL_S = 30;
const LOCK_RETRY_COUNT = 20;
const LOCK_RETRY_SLEEP_MS = 250;

/**
 * MD5-style 8-char hash for generating default Redis keys from token URLs.
 * Uses Web Crypto SHA-256 and takes the first 8 hex chars (collision probability
 * is negligible for the small number of providers a sidecar manages).
 */
async function urlHash(input: string): Promise<string> {
  const buf = await crypto.subtle.digest(
    "SHA-256",
    new TextEncoder().encode(input),
  );
  return Array.from(new Uint8Array(buf))
    .map((b) => b.toString(16).padStart(2, "0"))
    .join("")
    .slice(0, 8);
}

/**
 * OAuth2 client-credentials flow.
 *
 * Token is cached in tokenStorage with TTL = expires_in - 60.
 * A soft NX-style distributed lock prevents thundering-herd minting.
 *
 * Node (Redis) backend: uses SET NX EX for an atomic NX lock.
 * Workers (KV) backend: KV has no NX — a "lock" key is written with a short TTL;
 *   on miss, callers retry LOCK_RETRY_COUNT times before giving up. There is a
 *   narrow window where two Workers may both believe they hold the lock and both
 *   mint a token; the second write is a no-op (same value, same TTL) so the only
 *   cost is one extra token-endpoint round-trip.
 */
export class OAuth2ClientCredentialsAuth implements AuthModule {
  static readonly typeName = "oauth2_client_credentials";

  private constructor(
    private readonly deps: AuthDeps,
    private readonly clientId: string,
    private readonly clientSecret: string,
    private readonly tokenUrl: string,
    private readonly scope: string,
    private readonly audience: string,
    private readonly storageKey: string,
    private readonly lockKey: string,
    private readonly header: string,
    private readonly prefix: string,
    private readonly invalidateOn: Set<number>,
    private readonly ttlFallback: number,
  ) {}

  static async fromConfig(
    config: Record<string, unknown>,
    deps: AuthDeps,
  ): Promise<OAuth2ClientCredentialsAuth> {
    const cidEnv = config["client_id_env"];
    if (typeof cidEnv !== "string") {
      throw new Error(
        "OAuth2ClientCredentialsAuth: 'client_id_env' must be a string.",
      );
    }
    const csecEnv = config["client_secret_env"];
    if (typeof csecEnv !== "string") {
      throw new Error(
        "OAuth2ClientCredentialsAuth: 'client_secret_env' must be a string.",
      );
    }

    const clientId = deps.env[cidEnv];
    if (typeof clientId !== "string" || !clientId) {
      throw new Error(
        `OAuth2ClientCredentialsAuth: env var '${cidEnv}' is not set or empty.`,
      );
    }
    const clientSecret = deps.env[csecEnv];
    if (typeof clientSecret !== "string" || !clientSecret) {
      throw new Error(
        `OAuth2ClientCredentialsAuth: env var '${csecEnv}' is not set or empty.`,
      );
    }

    const tokenUrl = config["token_url"];
    if (typeof tokenUrl !== "string" || !tokenUrl) {
      throw new Error(
        "OAuth2ClientCredentialsAuth: 'token_url' must be a non-empty string.",
      );
    }

    const hash = await urlHash(tokenUrl);
    const defaultKey = `oauth2:${hash}`;
    const storageKey =
      typeof config["redis_key"] === "string"
        ? config["redis_key"]
        : defaultKey;
    const lockKey =
      typeof config["redis_lock_key"] === "string"
        ? config["redis_lock_key"]
        : `${storageKey}:lock`;

    return new OAuth2ClientCredentialsAuth(
      deps,
      clientId,
      clientSecret,
      tokenUrl,
      typeof config["scope"] === "string" ? config["scope"] : "",
      typeof config["audience"] === "string" ? config["audience"] : "",
      storageKey,
      lockKey,
      typeof config["header"] === "string" ? config["header"] : "Authorization",
      typeof config["prefix"] === "string" ? config["prefix"] : "Bearer ",
      invalidateOnFromConfig(config, [401]),
      typeof config["ttl_fallback"] === "number"
        ? config["ttl_fallback"]
        : 3600,
    );
  }

  private async fetchToken(): Promise<{ token: string; ttl: number }> {
    const body = new URLSearchParams({
      grant_type: "client_credentials",
      client_id: this.clientId,
      client_secret: this.clientSecret,
    });
    if (this.scope) body.set("scope", this.scope);
    if (this.audience) body.set("audience", this.audience);

    const resp = await fetch(this.tokenUrl, {
      method: "POST",
      headers: { "Content-Type": "application/x-www-form-urlencoded" },
      body: body.toString(),
      signal: AbortSignal.timeout(15_000),
    });

    if (!resp.ok) {
      throw new Error(
        `OAuth2ClientCredentialsAuth: token endpoint returned ${resp.status}.`,
      );
    }

    const data = (await resp.json()) as Record<string, unknown>;
    const accessToken = data["access_token"];
    if (typeof accessToken !== "string" || !accessToken) {
      throw new Error(
        "OAuth2ClientCredentialsAuth: response missing 'access_token'.",
      );
    }
    const expiresIn =
      typeof data["expires_in"] === "number"
        ? data["expires_in"]
        : this.ttlFallback;
    const ttl = Math.max(60, expiresIn - 60);
    return { token: accessToken, ttl };
  }

  async apply(_ctx: AuthContext): Promise<AuthResult> {
    const cached = await this.deps.tokenStorage.get(this.storageKey);
    if (cached) {
      return { headers: { [this.header]: `${this.prefix}${cached}` } };
    }

    // Attempt to acquire lock (SET NX equivalent).
    // On Redis: pttl of the lock key will be -2 (missing) → we write and proceed.
    // On KV: same pattern but non-atomic — see class docstring.
    const lockPttl = await this.deps.tokenStorage.pttl(this.lockKey);
    const hasLock = lockPttl === -2; // key is absent → we can take it

    if (!hasLock) {
      // Another worker is minting. Retry until token appears.
      for (let i = 0; i < LOCK_RETRY_COUNT; i++) {
        await sleep(LOCK_RETRY_SLEEP_MS);
        const retryVal = await this.deps.tokenStorage.get(this.storageKey);
        if (retryVal) {
          return { headers: { [this.header]: `${this.prefix}${retryVal}` } };
        }
      }
      throw new Error(
        `OAuth2ClientCredentialsAuth: timed out waiting for peer to fetch token (key=${JSON.stringify(this.storageKey)}).`,
      );
    }

    // Write lock with short TTL so it self-expires on crash.
    await this.deps.tokenStorage.set(this.lockKey, "1", LOCK_TTL_S);
    try {
      const { token, ttl } = await this.fetchToken();
      await this.deps.tokenStorage.set(this.storageKey, token, ttl);
      return { headers: { [this.header]: `${this.prefix}${token}` } };
    } finally {
      await this.deps.tokenStorage.delete(this.lockKey);
    }
  }

  async invalidate(): Promise<void> {
    await this.deps.tokenStorage.delete(this.storageKey);
  }

  isRejection(statusCode: number, _body: unknown): boolean {
    return this.invalidateOn.has(statusCode);
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
